pkg/microservice/aslan/core/templatestore/service/chart.go

```diff
@@ -0,0 +1,246 @@
+	s3Base := configbase.ObjectStorageChartTemplatePath(name)
+	if err = fs.PreloadFiles(name, localBase, s3Base, logger); err != nil {
+		return nil, err
+	}
+
+	base := filepath.Base(chart.Path)
+	file := filepath.Join(localBase, base, filePath, fileName)
+	fileContent, err := os.ReadFile(file)
+	if err != nil {
+		logger.Errorf("Failed to read file %s, err: %s", file, err)
+		return nil, err
+	}
+
+	return fileContent, nil
+}
+
+func AddChartTemplate(name string, args *fs.DownloadFromSourceArgs, logger *zap.SugaredLogger) error {
+	if mongodb.NewChartColl().Exist(name) {
+		return fmt.Errorf("a chart template with name %s is already existing", name)
+	}
+
+	sha1, err := processChartFromSource(name, args, logger)
+	if err != nil {
+		logger.Errorf("Failed to create chart %s, err: %s", name, err)
+		return err
+	}
+
+	return mongodb.NewChartColl().Create(&models.Chart{
+		Name:       name,
+		Owner:      args.Owner,
+		Repo:       args.Repo,
+		Path:       args.Path,
+		Branch:     args.Branch,
+		CodeHostID: args.CodehostID,
+		Sha1:       sha1,
+	})
+}
+
+func UpdateChartTemplate(name string, args *fs.DownloadFromSourceArgs, logger *zap.SugaredLogger) error {
+	chart, err := mongodb.NewChartColl().Get(name)
+	if err != nil {
+		logger.Errorf("Failed to get chart template %s.", name)
+		return err
+	}
+
+	sha1, err := processChartFromSource(name, args, logger)
+	if err != nil {
+		logger.Errorf("Failed to update chart %s, err: %s", name, err)
+		return err
+	}
+
+	if chart.Sha1 == sha1 {
+		logger.Debug("Chart %s has no changes, skip updating.", name)
+		return nil
+	}
+
+	return mongodb.NewChartColl().Update(&models.Chart{
+		Name:       name,
+		Owner:      args.Owner,
+		Repo:       args.Repo,
+		Path:       args.Path,
+		Branch:     args.Branch,
+		CodeHostID: args.CodehostID,
+		Sha1:       sha1,
+	})
+}
+
+func RemoveChartTemplate(name string, logger *zap.SugaredLogger) error {
+	err := mongodb.NewChartColl().Delete(name)
+	if err != nil {
+		logger.Errorf("Failed to delete chart template %s.", name)
+		return err
+	}
+
+	if err = fs.DeleteArchivedFileFromS3(name, configbase.ObjectStorageChartTemplatePath(name), logger); err != nil {
+		logger.Warnf("Failed to delete file %s, err: %s", name, err)
+	}
+
+	return nil
+}
+
+func processChartFromSource(name string, args *fs.DownloadFromSourceArgs, logger *zap.SugaredLogger) (string, error) {
+	tree, err := fs.DownloadFilesFromSource(args, func(a afero.Fs) (string, error) {
+		return "", nil
+	})
+
+	if err != nil {
+		logger.Errorf("Failed to download files with option %+v, err: %s", args, err)
+		return "", err
+	}
+
+	var sha1 string
+	if name == "" {
+		name = filepath.Base(args.Path)
+	}
+
+	var wg wait.Group
+	wg.Start(func() {
+		logger.Debug("Start to save and upload chart")
+		localBase := configbase.LocalChartTemplatePath(name)
+		s3Base := configbase.ObjectStorageChartTemplatePath(name)
+
+		err1 := fs.SaveAndUploadFiles(tree, name, localBase, s3Base, logger)
```

Comment:
1. You can also use err directly

Reply:
there is a bit difference if any err is nil in any of the goroutines